# a4 - Data Analysis
Fill in the below code cells as specified. Note that cells may utilize variables and functions defined in previous cells; we should be able to use the `Kernal > Restart & Clear All` menu item followed by `Cell > Run All` to execute your entire notebook and see the correct output.

## Part 1. Numbers
For this part of the assignment, you will analyze some numeric data (counts of library holdings) to investiate how the distribution of numbers in natural data sets obeys the counter-intuitive [Benford's Law](https://plus.maths.org/content/os/issue9/features/benford/index). 

<small>(This exercise was adapted from Steve Wolfman).</small>

Create a variable **`holdings_data`** which is a **list** of the contents of the **`data/libraryholdings.txt`** file included in the repository (each line in the file should be a single element in the list). You will need to open up the file and read its contents into a list. You should specify a _local path_ to the file (from this notebook's location).

In [1]:
with open('data/libraryholdings.txt', 'r') as f:
    for lines in f:
        holdings_data= f.readlines()
    print(holdings_data)




['(* collected by Christian Ayotte.                 *)\n', '(* Labels not available.                          *)\n', '\n', '12201\n', '600778\n', '14926\n', '37863\n', '14866\n', '9896\n', '8064\n', '9047\n', '3388\n', '21625\n', '8779\n', '7150\n', '9441\n', '10993\n', '7850\n', '7445\n', '276157\n', '13902\n', '13078\n', '8658\n', '22852\n', '21803\n', '17050\n', '34419\n', '3240\n', '106000\n', '8365\n', '74343\n', '55626\n', '65248\n', '15390\n', '11693\n', '8248\n', '14566\n', '11296\n', '11300\n', '17926\n', '2199\n', '42999\n', '7842\n', '28966\n', '11068\n', '2733\n', '11538\n', '29684\n', '57448\n', '11824\n', '34681\n', '10701\n', '4000\n', '13555\n', '7685\n', '6850\n', '11500\n', '46628\n', '57232\n', '21774\n', '45531\n', '8700\n', '6520\n', '10741\n', '8458\n', '43690\n', '11735\n', '18558\n', '3213\n', '4802\n', '6500\n', '4952\n', '9255\n', '4600\n', '1050\n', '7178\n', '3817\n', '991\n', '14064\n', '4200\n', '4680\n', '2566\n', '4830\n', '2781\n', '5884\n', '6564\n', '

Print out the first **ten** items from the `holdings_data` list, each on its own line. (Note that there may be extra line breaks that are included in the data items themselves).

In [2]:
#holdings_data[:10]

for i in range(10):
    print(holdings_data[i])

(* collected by Christian Ayotte.                 *)

(* Labels not available.                          *)



12201

600778

14926

37863

14866

9896

8064



Use the **slice operator (`:`)** to remove the "heading" and blank elements from the beginning of the data list, leaving only the list of numbers. The remaining values should continue to be stored (re-stored) in the `holdings_data` variable. Output the new first element in `holdings_data` to demonstrate that it is the first number in the data set.
- The values in the list _should_ be strings rather than an integers

In [3]:

#Remove heading lines:
with open('data/libraryholdings.txt') as f:
    for lines in f:
        holdings_data = f.readlines()[3:]
        
print(holdings_data)
        
# New first element :
print(holdings_data[0])
#String:
print(type(holdings_data[0]))

['12201\n', '600778\n', '14926\n', '37863\n', '14866\n', '9896\n', '8064\n', '9047\n', '3388\n', '21625\n', '8779\n', '7150\n', '9441\n', '10993\n', '7850\n', '7445\n', '276157\n', '13902\n', '13078\n', '8658\n', '22852\n', '21803\n', '17050\n', '34419\n', '3240\n', '106000\n', '8365\n', '74343\n', '55626\n', '65248\n', '15390\n', '11693\n', '8248\n', '14566\n', '11296\n', '11300\n', '17926\n', '2199\n', '42999\n', '7842\n', '28966\n', '11068\n', '2733\n', '11538\n', '29684\n', '57448\n', '11824\n', '34681\n', '10701\n', '4000\n', '13555\n', '7685\n', '6850\n', '11500\n', '46628\n', '57232\n', '21774\n', '45531\n', '8700\n', '6520\n', '10741\n', '8458\n', '43690\n', '11735\n', '18558\n', '3213\n', '4802\n', '6500\n', '4952\n', '9255\n', '4600\n', '1050\n', '7178\n', '3817\n', '991\n', '14064\n', '4200\n', '4680\n', '2566\n', '4830\n', '2781\n', '5884\n', '6564\n', '1770\n', '8150\n', '1200\n', '6274\n', '9209\n', '6730\n', '9736\n', '9577\n', '54347\n', '10750\n', '17744\n', '22453\n',

Create a variable **`lead_digit_counts`** that is a dictionary whose keys are _strings_ of each digit (`"0"`, `"1"`, `"2"`, etc.), and whose values are all the number `0`. You can do this directly or with a loop. Print out the variable after you create it.

In [4]:
lead_digit_counts={}
#All values are 0:
for i in range(0,10):
    lead_digit_counts[str(i)]=0
lead_digit_counts



{'0': 0,
 '1': 0,
 '2': 0,
 '3': 0,
 '4': 0,
 '5': 0,
 '6': 0,
 '7': 0,
 '8': 0,
 '9': 0}

Calculate the number of times each digit appears as the _first digit_ in a value of the `holdings_data` list, storing those counts in the `lead_digit_counts` dictionary.

In [5]:

for i in range(0,len(holdings_data)):
    for val in range(0,10):
        if(holdings_data[i][0]==str(val)):
            lead_digit_counts[str(val)]=lead_digit_counts[str(val)]+1

print(lead_digit_counts)

{'0': 0, '1': 3056, '2': 1606, '3': 1018, '4': 801, '5': 640, '6': 560, '7': 502, '8': 503, '9': 452}


Use a loop to print out each count in `lead_digit_counts` with the format:
```
X values have a leading digit of digit Y
```

In [6]:
for key,value in lead_digit_counts.items():
    print(value,"values have a leading digit of", key)

0 values have a leading digit of 0
3056 values have a leading digit of 1
1606 values have a leading digit of 2
1018 values have a leading digit of 3
801 values have a leading digit of 4
640 values have a leading digit of 5
560 values have a leading digit of 6
502 values have a leading digit of 7
503 values have a leading digit of 8
452 values have a leading digit of 9


Print the _percentage_ of values in the the library holdings data set that have a leading digit **`1`** (round to 2 decimal places). Is this value as predicted by Benford's law?

In [8]:
import math

Percentage_val = round(lead_digit_counts['1']/len(holdings_data),2)*100
print ("Percentage of words in the file that have a leading digit 1: ",Percentage_val)
        
# Benford's law = log((n+1)/n)
Ben_law = math.log10((1+1)/1)*100

print("Value predicted by Benford's law: ", Ben_law)


Percentage of words in the file that have a leading digit 1:  33.0
Value predicted by Benford's law:  30.10299956639812


***Extra credit challenge:*** Create a single variable `digit_position_counts` that contains the number of times that each digit 0 through 9 appears in _each_ position in the data set. E.g., a `1` appears in the 1st position 3056 times and in the second position 1005 times; a `2` appears in the 1st position 1606 times and in the second position 1044 times.

Use this variable to print a "table" of the percentage of the time each position contains each digit (e.g., the 1st digit is a `1` 33.44% of the time, a `2` 17.57% of the time, etc).

Note that for this extra challenge it is up to you to determine an appropriate data structure (e.g., how to combine dictionaries and lists and tuples) for representing this table. Be sure and include comments explaining your work.

Only attempt this problem once you have completed everything else!

## Part 2. Life Expectancy
For this part of the assignment, you'll work with data about the life expectancy (in years) for each country in the world in the years 1960 and 2013. Note that this can be really [fun](http://www.ted.com/talks/hans_rosling_shows_the_best_stats_you_ve_ever_seen.html) data!

The data is found in a [.csv](https://en.wikipedia.org/wiki/Comma-separated_values) file: a plain-text data format where each line represents a record (row) of data and where feature (column) is separated by a comma.

Read in the contents of the **`data/life_expectancy.csv`** data file, and use it to construct a **list** called **`life_expectancy_list`**. Each element in this list should be a **dictionary** (one for each row in the `csv` file) with the following keys and values:

- a key `'country'` whose value is the name of the country (as a string)
- a key `'le_1960'` whose value is the life expectancy in 1960 (as a float)
- a key `'le_2013'` whose value is the life expectancy in 2013 (as a float)

Thus the first record should look like:
```
{'country': 'Aruba', 'le_1960': 65.56936585, 'le_2013': 75.33217073}
```

You should use the **`csv`** module to read this file and break up each row into different values. See [the documentation](https://docs.python.org/3/library/csv.html) for an example of how to do this. Print out the _first row_ of your list as a demonstration that you've processed the data correctly.

In [10]:
import csv

life_expectancy_list=[]
with open('data/life_expectancy.csv') as csvfile:
    f = csv.reader(csvfile)
    for val in f:
            dicts={}
            dicts['country']=val[0]
            dicts['le_1960']=val[-2]
            dicts['le_2013']=val[-1]
            life_expectancy_list.append(dicts)
life_expectancy_list = life_expectancy_list[1:]  

#Print first record:
print(life_expectancy_list[0])



{'country': 'Aruba', 'le_1960': '65.56936585', 'le_2013': '75.33217073'}


Add another item to each dictionary in the `life_expectancy_list` whose **key** is `change` and whose **value** is the change in life expectancy from 1960 to 2013.

In [11]:
for val in life_expectancy_list:
    val['change'] = (float(val['le_2013'])) - (float(val['le_1960'])) 
print(life_expectancy_list)

[{'country': 'Aruba', 'le_1960': '65.56936585', 'le_2013': '75.33217073', 'change': 9.762804880000004}, {'country': 'Afghanistan', 'le_1960': '31.58004878', 'le_2013': '60.93141463', 'change': 29.35136585}, {'country': 'Angola', 'le_1960': '32.98482927', 'le_2013': '51.86617073', 'change': 18.88134146}, {'country': 'Albania', 'le_1960': '62.25436585', 'le_2013': '77.5372439', 'change': 15.282878050000008}, {'country': 'United Arab Emirates', 'le_1960': '52.24321951', 'le_2013': '77.13129268', 'change': 24.88807317}, {'country': 'Argentina', 'le_1960': '65.21553659', 'le_2013': '76.18729268', 'change': 10.97175609}, {'country': 'Armenia', 'le_1960': '65.86346341', 'le_2013': '74.5407561', 'change': 8.677292690000002}, {'country': 'Antigua and Barbuda', 'le_1960': '61.78273171', 'le_2013': '75.82929268', 'change': 14.046560969999994}, {'country': 'Australia', 'le_1960': '70.81707317', 'le_2013': '82.19756098', 'change': 11.380487810000005}, {'country': 'Austria', 'le_1960': '68.58560976'

Create a variable **`num_small_gain`** that stores the **number of countries** whose life expectancy did not improve by 5 years or more between 1960 and 2013. This will include counties whose life expectancy has worsened. Print out this variable.

In [12]:
num_small_gain =0
#change in life expectancy<5 years
for val in life_expectancy_list:
    if int(val['change']) < 5:
        num_small_gain = num_small_gain+1
print("Number of countries: ", num_small_gain)

Number of countries:  7


Create a variable **`most_improved`** that is the **name of the country** with the largest gain in life expectancy (between 1960 and 2013). Print out this variable.

In [15]:
maximum_val = life_expectancy_list[0]['change']
most_improved = life_expectancy_list[0]['country']
for val in life_expectancy_list:
    if val['change'] > maximum_val:
        maximum_val = val['change']
        most_improved = val['country']
        
print("Name of the country with the largest gain in life expectancy (between 1960 and 2013): " ,most_improved)
print("Value is: ", maximum_val)


Name of the country with the largest gain in life expectancy (between 1960 and 2013):  Maldives
Value is:  42.07575609


Define a function **`compare_country_le()`** that takes in the names of _two_ countries, and returns a **tuple** containing the following information:
- the name of the country with the greater life expectancy,
- the life expectancy in 2013 of that country
- the difference between the life expectancies in 2013

Use your function to print the comparison between the life expectancies of the _United States_ and _Cuba_.  

In [17]:
def compare_country_le(country1,country2):
    
    for val1 in life_expectancy_list:
        for val2 in life_expectancy_list:
            if(val1['country'] == country1):
                country_variable1 = (val1['le_2013'], country1)
            if (val2['country'] == country2):
                country_variable2 = (val2['le_2013'], country2)
    difference = abs(float(country_variable1[0]) - float(country_variable2[0]))
    if max(country_variable1[0],country_variable2[0]) == country_variable1[0]:
        country_greater_le = country_variable1[1]
    else:
        country_greater_le = country_variable2[1]
    maximum_le = (max(country_variable1[0],country_variable2[0]),country_greater_le, difference)
    return maximum_le

#call function:
compare_country_le('United States','Cuba')

('79.23926829', 'Cuba', 0.39780487999999536)

## Part 3. Readability
For this part of the assignment, you will calculate the [readability](https://en.wikipedia.org/wiki/Readability) of a text document using the [Dale-Chall Readability Formula](http://www.readabilityformulas.com/new-dale-chall-readability-formula.php). This method determines how "easy" it is to read a particular (English) document by considering the length of sentences and how many of the words used are "easy" to understand (based on a pre-defined list of "easy" words).

Splitting real-world text documents into words and sentences is non-trivial (English is hard!). To make this easier, you should use the [Natural Language Toolkit (nltk)](http://www.nltk.org/index.html) module. This module is included with Anacaonda, but does require some additional data source files to be installed on your computer. You _should_ be able to do this by running the below cell (you only need to run it once).

In [18]:
from nltk import download
download('punkt')
download('wordnet')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Neha\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Neha\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

You will also need to load the list of "easy" words into memory. This list can be found in the **`data/dale-chall.txt`** file. Open this file and read its entire contents into a **list** variable (e.g., `easy_words_list`), where each element in the list is a single line (word) in the file.

In [19]:
with open('data/dale-chall.txt', 'r') as f:
    for lines in f:
        easy_words_list= f.readlines()
        print(easy_words_list)

['able\n', 'aboard\n', 'about\n', 'above\n', 'absent\n', 'accept\n', 'accident\n', 'account\n', 'ache\n', 'aching\n', 'acorn\n', 'acre\n', 'across\n', 'act\n', 'acts\n', 'add\n', 'address\n', 'admire\n', 'adventure\n', 'afar\n', 'afraid\n', 'after\n', 'afternoon\n', 'afterward\n', 'afterwards\n', 'again\n', 'against\n', 'age\n', 'aged\n', 'ago\n', 'agree\n', 'ah\n', 'ahead\n', 'aid\n', 'aim\n', 'air\n', 'airfield\n', 'airplane\n', 'airport\n', 'airship\n', 'airy\n', 'alarm\n', 'alike\n', 'alive\n', 'all\n', 'alley\n', 'alligator\n', 'allow\n', 'almost\n', 'alone\n', 'along\n', 'aloud\n', 'already\n', 'also\n', 'always\n', 'am\n', 'america\n', 'american\n', 'among\n', 'amount\n', 'an\n', 'and\n', 'angel\n', 'anger\n', 'angry\n', 'animal\n', 'another\n', 'answer\n', 'ant\n', 'any\n', 'anybody\n', 'anyhow\n', 'anyone\n', 'anything\n', 'anyway\n', 'anywhere\n', 'apart\n', 'apartment\n', 'ape\n', 'apiece\n', 'appear\n', 'apple\n', 'april\n', 'apron\n', 'are\n', "aren't\n", 'arise\n', 'arith

In order to "look up" easy words, convert the easy words list into a **dictionary** (e.g., `easy_words_dict`), where each **key** is a word, and each **value** is `True` (that the word is in the list).
- Make sure you do not include newline characters in your keys!

In [20]:
easy_words_dict={}
#Each value is true:
for value in easy_words_list:
    easy_words_dict = {k : 'TRUE' for k in (lines.strip('\n') for lines in easy_words_list)}
print(easy_words_dict)

{'able': 'TRUE', 'aboard': 'TRUE', 'about': 'TRUE', 'above': 'TRUE', 'absent': 'TRUE', 'accept': 'TRUE', 'accident': 'TRUE', 'account': 'TRUE', 'ache': 'TRUE', 'aching': 'TRUE', 'acorn': 'TRUE', 'acre': 'TRUE', 'across': 'TRUE', 'act': 'TRUE', 'acts': 'TRUE', 'add': 'TRUE', 'address': 'TRUE', 'admire': 'TRUE', 'adventure': 'TRUE', 'afar': 'TRUE', 'afraid': 'TRUE', 'after': 'TRUE', 'afternoon': 'TRUE', 'afterward': 'TRUE', 'afterwards': 'TRUE', 'again': 'TRUE', 'against': 'TRUE', 'age': 'TRUE', 'aged': 'TRUE', 'ago': 'TRUE', 'agree': 'TRUE', 'ah': 'TRUE', 'ahead': 'TRUE', 'aid': 'TRUE', 'aim': 'TRUE', 'air': 'TRUE', 'airfield': 'TRUE', 'airplane': 'TRUE', 'airport': 'TRUE', 'airship': 'TRUE', 'airy': 'TRUE', 'alarm': 'TRUE', 'alike': 'TRUE', 'alive': 'TRUE', 'all': 'TRUE', 'alley': 'TRUE', 'alligator': 'TRUE', 'allow': 'TRUE', 'almost': 'TRUE', 'alone': 'TRUE', 'along': 'TRUE', 'aloud': 'TRUE', 'already': 'TRUE', 'also': 'TRUE', 'always': 'TRUE', 'am': 'TRUE', 'america': 'TRUE', 'americ

Additionally, define a dictionary **`readability_grade_dict`** to use for looking up the "grade level" associated with a readability score (see [this table](https://en.wikipedia.org/wiki/Dale%E2%80%93Chall_readability_formula#Formula)). This dictionary should have **keys** that are ___tuples___ giving the range of score for a particular grade (e.g., `(5.0, 5.9)`), and **values** that are ___strings___ representing the grade (e.g., `"5th or 6th grade"`). 

In [21]:
#readability_grade_dict

keys = ['(0, 4.9)', '(5.0, 5.9)', '(6.0, 6.9)','(7.0, 7.9)','(8.0, 8.9)','(9.0, 9.9)']
values = ['easily understood by an average 4th-grade studen or lower', 'easily understood by an average 5th or 6th-grade student', 'easily understood by an average 7th or 8th-grade student','easily understood by an average 9th or 10th-grade student','easily understood by an average 11th or 12th-grade student','easily understood by an average 13th to 15th-grade (college) student']
readability_grade_dict=dict(zip(keys, values)) 

Define a function **`print_grade()`** that takes in a readability score (a number greater than or equal to 0), and **prints** a string representing the grade associated with that score (from your `readability_grade_dict` dictionary).
- _Hint:_ loop through the items in the dictionary and determine which "tuple" key has elements that the score falls between. Be sure and round to the nearest decimal).

In [22]:
import math

def print_grade(score):
    for key,value in readability_grade_dict.items():
        if (str(math.floor(score))) in key:
            print(value)

print_grade(4.5)

easily understood by an average 4th-grade studen or lower


Now to calculate the readability scores! Define a function **`count_sentences()`** that counts the number of sentences in a string. Use the [sent_tokenize()](http://www.nltk.org/api/nltk.tokenize.html#nltk.tokenize.sent_tokenize) function from the `nltk.tokenize` module to break up a string into sentences (this is like the string `split()` function, but it splits into sentences rather than dividing by spaces).
- For help and an example, see [this guide](http://textminingonline.com/dive-into-nltk-part-ii-sentence-tokenize-and-word-tokenize).
- You do not need to do any extra processing beyond that provided by the `sent_tokenize()` function.
- Test your function on a simple pair or trio of sentences!

In [23]:
def count_sentences(text):
    
    from nltk.tokenize import sent_tokenize
    sent_tokenize_list = sent_tokenize(text)
    print(len(sent_tokenize_list))
    print(sent_tokenize_list)
    return (len(sent_tokenize_list))
count_sentences("a sent tokenize test. this is sent two. is this sent three? sent 4 is cool! Now it’s your turn.")

5
['a sent tokenize test.', 'this is sent two.', 'is this sent three?', 'sent 4 is cool!', 'Now it’s your turn.']


5

Define a function **`extract_words()`** that takes in a string and returns a _list_ of all of the words in that string. Use the [word_tokenize()](http://www.nltk.org/api/nltk.tokenize.html#nltk.tokenize.word_tokenize) function from the `nltk.tokenize` module to break up the string into words.
- The `nltk` tokenizer includes each punctuation character (e.g., commas, periods) as individual "words". Your list should not include these items. You can use a string method to determine whether or not the word starts with a punctuation symbol, and if so exclude it. _Hint_ think about keeping good words, rather than throwing away the bad! Note that you do not need to do any special consideration for contractions or other words that include their own punctuation.
- Test your function on a simple sentence (with punctuation!).

In [25]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.tokenize import RegexpTokenizer
import string
from nltk.tokenize import WordPunctTokenizer


def extract_words(string):
    words = RegexpTokenizer(r'\w+').tokenize(string)
    return words

extract_words("I will not do this now, because I'm so tired.  Please give me some time. @ sd  4 232")
    

['I',
 'will',
 'not',
 'do',
 'this',
 'now',
 'because',
 'I',
 'm',
 'so',
 'tired',
 'Please',
 'give',
 'me',
 'some',
 'time',
 'sd',
 '4',
 '232']

In [ ]:
#Alternate method:
# text = "'I wonder how many miles I've fallen by this time?' she said aloud."
# import string
# words = [x.strip(string.punctuation) for x in text.split()]
# words

Define a function **`count_easy_words()`** that takes in a _list_ of words as an argument and returns the number of words that are "easy".

- Your function should look up each word in the `easy_words_dict` you defined earlier. _Do not look up words in the list_ (the dictionary is much faster!). Be careful to look up lowercase versions of the word.

- Your function should handle detecting different parts of speech (e.g., plurals, different verb conjugations, etc.). You can do this by using the **`WordNetLemmatizer()`** function from the `nltk.stem.wordnet` module&mdash;which produces a "lemmatizer" object. You can call the **`lemmatize()`** method on this object to reduce a word to its "base" form. See [this example](https://pythonprogramming.net/lemmatizing-nltk-tutorial/). Note that you should reduce words to both their basic noun AND verb forms (you will need to call the function twice: once with `'n'` (noun) and once with `'v'` (verb) as the second argument!)

- You can test your function on the word list: `['My','words','spoken','have','consequences']`, which should have 4 of the 5 words considered easy (not "consequences").

In [33]:
from nltk.stem import WordNetLemmatizer

def count_easy_words(wordlist, form_type):
    easywords_list =[]
    lemmatize_list =[]
    for i in range(len(wordlist)):
        lemmatize_list.append(WordNetLemmatizer().lemmatize((wordlist[i]).lower(), form_type))
        if lemmatize_list[i] in easy_words_dict.keys():
            easywords_list.append(lemmatize_list[i])
    return len(easywords_list)

#Enter noun or verb as second argument:
count_easy_words(['My','words','spoken','have','consequences'],'v')
count_easy_words(['My','words','spoken','have','consequences'],'n')

3

Define a function **`calc_readability_score()`** that takes in a string of text and returns a readability "score" for the test based on the [Dale-Chall readability formula](https://en.wikipedia.org/wiki/Dale%E2%80%93Chall_readability_formula#Formula). Call your previous functions to calculate the number of sentences, total words, and number of difficult (not easy) words.
- Don't forget to adjust the score if the text is more than 5% difficult words!

In [40]:
def calc_readability_score(string):
    count_of_sentences = count_sentences(string)
    print("no of sentences-",count_of_sentences)
    tokenize_words = extract_words(string)
    len_tokenize_words=len(tokenize_words)
#     print(len_tokenize_words)
#     print(tokenize_words)
    number_of_easy_words = count_easy_words(tokenize_words,'v')
    print("no of easy words",number_of_easy_words)
    not_easy_words = len(tokenize_words) - number_of_easy_words
    print(not_easy_words)
    raw_score = (0.1579*((not_easy_words/len_tokenize_words))*100)+ (0.0456*(number_of_easy_words/count_of_sentences))
    print(raw_score)
    percentage_diff = (not_easy_words/len(tokenize_words))*100
    adjusted_score =0
    #Adjust raw score:
    if percentage_diff > 5:
        adjusted_score = raw_score + 3.6365
        return(adjusted_score)
    else:
        return(raw_score)


calc_readability_score("This is my address.I live here")


1
['This is my address.I live here']
no of sentences- 1
no of easy words 7
0
0.31920000000000004


0.31920000000000004

Read in the text of the `data/alice.txt` file (the full text of Alice in Wonderland) _as a single string_. 

In [36]:
with open('data/alice.txt',encoding="utf8") as alice:
    alice_data = alice.read()
    print(alice_data)
    

ALICE’S ADVENTURES IN WONDERLAND

Lewis Carroll

THE MILLENNIUM FULCRUM EDITION 3.0




CHAPTER I. Down the Rabbit-Hole

Alice was beginning to get very tired of sitting by her sister on the
bank, and of having nothing to do: once or twice she had peeped into the
book her sister was reading, but it had no pictures or conversations in
it, ‘and what is the use of a book,’ thought Alice ‘without pictures or
conversations?’

So she was considering in her own mind (as well as she could, for the
hot day made her feel very sleepy and stupid), whether the pleasure
of making a daisy-chain would be worth the trouble of getting up and
picking the daisies, when suddenly a White Rabbit with pink eyes ran
close by her.

There was nothing so VERY remarkable in that; nor did Alice think it so
VERY much out of the way to hear the Rabbit say to itself, ‘Oh dear!
Oh dear! I shall be late!’ (when she thought it over afterwards, it
occurred to her that she ought to have wondered at this, but at the time
i

Calculate the readability score for the `alice.txt` file and print it out. Then print out the reading grade associated with that score. Use your previously-defined functions!
- For testing, note that my calculations show `alice.txt` has 977 sentences and 27198 words, of which 3610 are difficult. This leads to a readability score of ~7.113.

In [45]:
#Readability score 
alice_score = calc_readability_score(alice_data)
print("Readibility score: " ,alice_score)

# Grade
grade=print_grade(alice_score)


#number of sentences
print('The number of sentences : ', count_sentences(alice_data))

977
['\ufeffALICE’S ADVENTURES IN WONDERLAND\n\nLewis Carroll\n\nTHE MILLENNIUM FULCRUM EDITION 3.0\n\n\n\n\nCHAPTER I.', 'Down the Rabbit-Hole\n\nAlice was beginning to get very tired of sitting by her sister on the\nbank, and of having nothing to do: once or twice she had peeped into the\nbook her sister was reading, but it had no pictures or conversations in\nit, ‘and what is the use of a book,’ thought Alice ‘without pictures or\nconversations?’\n\nSo she was considering in her own mind (as well as she could, for the\nhot day made her feel very sleepy and stupid), whether the pleasure\nof making a daisy-chain would be worth the trouble of getting up and\npicking the daisies, when suddenly a White Rabbit with pink eyes ran\nclose by her.', 'There was nothing so VERY remarkable in that; nor did Alice think it so\nVERY much out of the way to hear the Rabbit say to itself, ‘Oh dear!', 'Oh dear!', 'I shall be late!’ (when she thought it over afterwards, it\noccurred to her that she ough

no of easy words 23169
4174
3.4917741756482186
Readibility score:  7.128274175648219
easily understood by an average 9th or 10th-grade student
977
['\ufeffALICE’S ADVENTURES IN WONDERLAND\n\nLewis Carroll\n\nTHE MILLENNIUM FULCRUM EDITION 3.0\n\n\n\n\nCHAPTER I.', 'Down the Rabbit-Hole\n\nAlice was beginning to get very tired of sitting by her sister on the\nbank, and of having nothing to do: once or twice she had peeped into the\nbook her sister was reading, but it had no pictures or conversations in\nit, ‘and what is the use of a book,’ thought Alice ‘without pictures or\nconversations?’\n\nSo she was considering in her own mind (as well as she could, for the\nhot day made her feel very sleepy and stupid), whether the pleasure\nof making a daisy-chain would be worth the trouble of getting up and\npicking the daisies, when suddenly a White Rabbit with pink eyes ran\nclose by her.', 'There was nothing so VERY remarkable in that; nor did Alice think it so\nVERY much out of the way to he

_Note that this result may not be an especially accurate model of a text's readability&mdash;after all, it's just based on a simple estimation!_